src/util.rs

```diff
@@ -37,30 +37,33 @@ pub(crate) mod ptr {
     /// `Ptr<'a, T>` is [covariant] in `'a` and `T`.
     ///
     /// [covariant]: https://doc.rust-lang.org/reference/subtyping.html
-    pub struct Ptr<'a, T: 'a + ?Sized> {
-        // INVARIANTS:
-        // 1. `ptr` is derived from some valid Rust allocation, `A`
-        // 2. `ptr` has the same provenance as `A`
-        // 3. `ptr` addresses a byte range which is entirely contained in `A`
-        // 4. `ptr` addresses a byte range whose length fits in an `isize`
-        // 5. `ptr` addresses a byte range which does not wrap around the address
-        //     space
-        // 6. `ptr` is validly-aligned for `T`
-        // 7. `A` is guaranteed to live for at least `'a`
-        // 8. `T: 'a`
+    pub struct Ptr<'a, T: 'a + ?Sized, const ASSUME_ALIGNED: bool> {
+        /// INVARIANTS:
+        /// 1. `ptr` is derived from some valid Rust allocation, `A`
+        /// 2. `ptr` has the same provenance as `A`
+        /// 3. `ptr` addresses a byte range which is entirely contained in `A`
+        /// 4. `ptr` addresses a byte range whose length fits in an `isize`
+        /// 5. `ptr` addresses a byte range which does not wrap around the address
+        ///     space
+        /// 6. If `ASSUME_ALIGNED`, `ptr` is validly-aligned for `T`
+        /// 7. `A` is guaranteed to live for at least `'a`
+        /// 8. `T: 'a`
         ptr: NonNull<T>,
         _lifetime: PhantomData<&'a ()>,
     }
 
-    impl<'a, T: ?Sized> Copy for Ptr<'a, T> {}
-    impl<'a, T: ?Sized> Clone for Ptr<'a, T> {
+    /// The argument to `TryFromBytes::is_bit_valid`.
+    pub type MaybeValid<'a, T> = Ptr<'a, T, false>;
+
+    impl<'a, T: ?Sized, const ASSUME_ALIGNED: bool> Copy for Ptr<'a, T, ASSUME_ALIGNED> {}
+    impl<'a, T: ?Sized, const ASSUME_ALIGNED: bool> Clone for Ptr<'a, T, ASSUME_ALIGNED> {
         #[inline]
         fn clone(&self) -> Self {
             *self
```

Comment:
While you're here, safety comment on this? We should have had one from the beginning.